### PR TITLE
[15.0][ADD] project_close_forecast_date

### PR DESCRIPTION
**Pull request by ChristianSantamaria**
Adds new field forecast date in purchase order line.

@dalonsod Can u review this addon?

### CODE REVIEW

File: project_close_forecast_date/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    "name": "Project Close Forecast Date",
+    "summary": """
+        Adds new field forecast date in purchase order line.
+    """,
+    "author": "Solvos",
+    "license": "LGPL-3",
+    "version": "15.0.1.0.0",
+    "category": "Project",
+    "website": "https://github.com/solvosci/slv-project",
+    "depends": ["project_purchase"],
```
Comment (dalonsod) [inline project_close_forecast_date/__manifest__.py]: This dependency is actually unused (`project_purchase` only adds a soft link between purchases and projects, nothing that seems to be used here), and stock dependency (due to `stock_picking`) is missing, then better try this:
```suggestion
    "depends": ["project", "purchase_stock"],

```

File: project_close_forecast_date/models/project_project.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+
+
+class ProjectProject(models.Model):
+    _inherit = 'project.project'
+
+    def write(self, vals):
+        if 'last_close_date' in vals and vals['last_close_date']:
+            order_line_ids = self.env['purchase.order.line'].search([
```
Comment (dalonsod) [inline project_close_forecast_date/models/project_project.py]: Have you check every user with project access can read every purchase order line?

File: project_close_forecast_date/models/project_project.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+                ('account_analytic_id', 'in', self.analytic_account_id.ids)
+            ])
+            date = datetime.strptime(vals['last_close_date'], '%Y-%m-%d').date()
```
Comment (dalonsod) [inline project_close_forecast_date/models/project_project.py]: we could skip `strptime()` process first calling to `super().write()` and later inspecting `self[0].last_close_date.date()`, is it right?

File: project_close_forecast_date/models/project_project.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+from odoo import models
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+
+
+class ProjectProject(models.Model):
+    _inherit = 'project.project'
+
+    def write(self, vals):
+        if 'last_close_date' in vals and vals['last_close_date']:
```
Comment (dalonsod) [inline project_close_forecast_date/models/project_project.py:13]: Ouch!!! Please check, but `last_close_date` is a non stardard field for `project.project` module, it's added by another addon, could you check?

File: project_close_forecast_date/models/purchase_order_line.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, fields, models
+from pytz import timezone
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = 'purchase.order.line'
+    
+    forecast_date = fields.Date()
```
Comment (dalonsod) [inline project_close_forecast_date/models/purchase_order_line.py]: Should this field be readonly/computed, or users can manually change it?

File: project_close_forecast_date/views/purchase_order_views.xml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record id="purchase_order_form" model="ir.ui.view">
+        <field name="name">purchase.order.form</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//tree/field[@name='date_planned']" position="after">
+                <field name="forecast_date" />
```
Comment (dalonsod) [inline project_close_forecast_date/views/purchase_order_views.xml:10]: If this field can be manually set by users, is there any state that should lock it, like other fields? And should it be visible in those cases for every user role?

File: project_close_forecast_date/models/purchase_order.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+
+class PurchaseOrder(models.Model):
+    _inherit = 'purchase.order'
+
+    @api.onchange('date_planned')
+    def _onchange_date_planned(self):
+        if self.order_line:
+            tz = timezone(self.env.user.tz or self.company_id.partner_id.tz or 'UTC')
+            self.order_line.forecast_date = self.date_planned.astimezone(tz).date()
```
Comment (dalonsod) [inline project_close_forecast_date/models/purchase_order.py]: This process is repeated below, then should be moved to an order line method, in order to not be repeated. I suspect that there's an Odoo base method that makes such conversion, if I find it, I'll tell you

**Review by dalonsod — CHANGES_REQUESTED** [review]
See comments. Please pay attention at dependencies, security access and readonly considerations

**Comment by ChristianSantamaria** [comment]
@dalonsod Updated Test enviroment

**Review by dalonsod — CHANGES_REQUESTED** [review]
This code still needs some refactoring an adjustments:
* `last_close_date` change in projects and POs security. When I noted that a Project User can't access POLs by default (if it's not a Purchase User as well), I wasn't saying that changing project last close date was a forbidden operation. I simply said that an error would be raised if we don't prevent that situation. I think that such operation should be actually enabled, so an ACL/record rule for POLs should be added or a `sudo()` when reading and writing POLs should be coded. IMO the last one is the right one (no matter who wants to change this date, later operations must be performed).
* UTC code for determining date from datetime is not duplicated, but wrote three times! Please unify it in a single method for e.g. purchase order line.
* Could you check this? When PO `date_planned` changes, every `date_planned` for its POLs are changed too, with this new value. So onchange management seems to be only required at POL level. This should work: making `forecast_date` computed stored. Compute method should only depends on POL `date_planned` field AND linked stock picking `scheduled_date` (so every `onchange` method could be removed, and `write` for picking too).

**Comment by ChristianSantamaria** [comment]
@dalonsod Updated

File: project_close_forecast_date/models/purchase_order_line.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+from odoo import _, api, fields, models
+from pytz import timezone
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = 'purchase.order.line'
+
+    forecast_date = fields.Date(compute='_compute_forecast_date', store=True, readonly=True)
+
+    @api.depends('date_planned')
```
Comment (dalonsod) [inline project_close_forecast_date/models/purchase_order_line.py:12]: If this dependency is added, `write()` stuff for `stock.picking` should become unneeded? Please check (when picking is created and when picking is modified):
```suggestion
    @api.depends('date_planned', 'move_ids.picking_id.scheduled_date') 
```

**Comment by ChristianSantamaria** [comment]
When change "last_close_date" in project update purchase.order.line while order_id.invoice_status != invoiced
@dalonsod can u check?

File: project_close_forecast_date/models/project_project.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+class ProjectProject(models.Model):
+    _inherit = 'project.project'
+
+    def write(self, vals):
+        res = super().write(vals)
+        if 'last_close_date' in vals and vals['last_close_date']:
+            order_line_ids = self.env['purchase.order.line'].sudo().search([
+                ('account_analytic_id', 'in', self.analytic_account_id.ids),
+                ('order_id.invoice_status', '!=', 'invoiced')
```
Comment (dalonsod) [inline project_close_forecast_date/models/project_project.py]: Well.. we might exclude cancelled lines IMO:
```suggestion
                ('order_id.invoice_status', '!=', 'invoiced'),
                ('state', '!=', 'cancel'),
```

File: project_close_forecast_date/models/purchase_order_line.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+    forecast_date = fields.Date(compute='_compute_forecast_date', store=True, readonly=True)
+
+    @api.depends('date_planned') 
+    def _compute_forecast_date(self):
+        for record in self:
+            record.forecast_date = record.get_date_tz(record.date_planned)
+
+    def get_date_tz(self, date):
+        tz = timezone(self.env.user.tz or self.company_id.partner_id.tz or 'UTC')
```
Comment (dalonsod) [inline project_close_forecast_date/models/purchase_order_line.py]: At last, I've found this:

https://github.com/odoo/odoo/blob/6708408c9d2f8137271809962f0d39ee38a0a24b/odoo/fields.py#L1887-L1909

Such function could be called like .e.g. `fields.Date.context_today(self, timestamp=date)`, could you try?

**Review by dalonsod — COMMENTED** [review]
See very last comments, I promise!

**Comment by ChristianSantamaria** [comment]
@dalonsod and now?

**Review by dalonsod — APPROVED** [review]
POL `forecast_date` will be initialized with `date_planned`.

:+1:

**Pull request merged**